core/src/main/scala/cats/derived/contravariant.scala

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright (c) 2015 Miles Sabin
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package cats
+package derived
+
+import shapeless._
+import util.VersionSpecific.OrElse
+
+import scala.annotation.implicitNotFound
+
+@implicitNotFound("Could not derive an instance of Contravariant[${F}]")
+trait MkContravariant[F[_]] extends Contravariant[F] {
+  def safeContramap[A, B](fa: F[A])(f: B => Eval[A]): Eval[F[B]]
+  def contramap[A, B](fa: F[A])(f: B => A): F[B] = safeContramap(fa)(f andThen Eval.now).value
+}
+object MkContravariant extends MkContravariantDerivation {
+  def apply[F[_]](implicit F: MkContravariant[F]): MkContravariant[F] = F
+}
+private[derived] abstract class MkContravariantDerivation extends MkContravariantNested {
+  implicit val mkContraHNil: MkContravariant[Const[HNil]#λ] = mkContraConst
+  implicit val mkContraCNil: MkContravariant[Const[CNil]#λ] = mkContraConst
+
+  implicit def mkContraConst[T]: MkContravariant[Const[T]#λ] =
+    new MkContravariant[Const[T]#λ] {
+      def safeContramap[A, B](t: T)(f: B => Eval[A]): Eval[T] = Eval.now(t)
+    }
+}
+
+private[derived] abstract class MkContravariantNested extends MkContravariantCons {
+  protected type FunctorOrMk[F[_]] = Functor[F] OrElse MkFunctor[F]
+  protected def mkSafeMap[F[_], A, B](F: FunctorOrMk[F])(fa: F[A])(f: A => Eval[B]): Eval[F[B]] =
+    F.unify match {
+      case mk: MkFunctor[F] => mk.safeMap(fa)(f)
+      case p => Eval.later(p.map(fa)(f(_).value))
+    }
```

Comment:
This is duplication, do we have a better way to reuse stuff across derivations?

Reply:
Yes, unfortunately you will see some code duplication, we haven't found a good way to get rid of it.

But in this case I would say we are trying to do too much. Kittens already tries to derive as much as it can but only for a single typeclass. E.g. when the user writes `semi.contravariant[F]` we should not try to derive a `Functor`, even just as an intermediate derivation.

Anyway forgive my rambling, what I'm trying to say is that `MkContravariant` shouldn't depend on `MkFunctor`. Instead just adding a `Functor` implicit where necessary is enough.

Reply:
Nice, ok that makes sense. I'll switch to that.